contracts/tg4-stake/src/contract.rs

```diff
@@ -2256,8 +2320,100 @@ mod tests {
             assert_eq!(sends, expected_transfers);
         }
 
+        /// Helper for asserting if expected undelegates occurred in response. Panics if any non
+        /// `TgradeMsg::Undelegate` occurred, or undelegates are different than expected.
+        ///
+        /// Transfers are passed in form of pairs `(addr, amount)`, as for all tests in this module
+        /// the expected denom is fixed.
+        #[track_caller]
+        fn assert_undelegates(response: Response, mut expected_transfers: Vec<(&str, u128)>) {
```

Comment:
Why do we want these kinds of tricks (which are just difficult to reason about and are additional logic in tests) instead of using multitest to test messaging? I know multitest natively doesn't have implementation for custom messages, but it is a good time to create a custom message handler for tgrade and use it for testing if particular actions were performed on chain.

Reply:
This is just building up on previous work. I wouldn't change the entire test setup now.